internal/concurrency/concurrency.go

```diff
@@ -0,0 +1,81 @@
+package concurrency
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"time"
+
+	"github.com/gofrs/flock"
+	"github.com/speakeasy-api/speakeasy/internal/log"
+	"github.com/speakeasy-api/speakeasy/internal/singleton"
+)
+
+// Package concurrency provides utilities for inter-process synchronization.
+
+// InterProcessMutex provides file-based mutual exclusion between processes.
+// The lock is automatically released if the holding process dies.
+//
+// See:
+//   - Linux: https://linux.die.net/man/2/flock
+//   - Windows: https://docs.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-lockfileex
+type InterProcessMutex struct {
+	Opts
+	mu *flock.Flock
+}
+
+type Opts struct {
+	Name           string
+	LockRetryDelay time.Duration
+}
+
+func DefaultOpts() Opts {
+	return Opts{Name: "speakeasy.lock", LockRetryDelay: 10 * time.Second}
+}
+
+func new(o Opts) *InterProcessMutex {
+	mu := flock.New(filepath.Join(os.TempDir(), o.Name))
+
+	return &InterProcessMutex{mu: mu}
+}
+
+// NewIPMutexWithOpts creates a new inter-process mutex with a custom config.
+var NewIPMutexWithOpts = singleton.NewWithOpts(func(o Opts) *InterProcessMutex {
+	return new(o)
+})
+
+// NewIPMutex creates a new inter-process mutex with the default config.
+var NewIPMutex = singleton.New(func() *InterProcessMutex {
+	return new(DefaultOpts())
+})
```

Comment:
```suggestion
var GetCLIUpgradeMutex = singleton.New(func() *InterProcessMutex {
	return new({ Name: "speakeasy-update-cli.lock", Timeout: time.Minute })
})
```

Reply:
Naming wise, `IP` makes me think ip address. I think interprocess / file system lock / os lock is good to be referenced in doc comments. This particular singleton should be specific to upgrading / pinning / installation of the CLI rather than NewIPMutex, we could have other use cases for the IP mutex